File: 2020/py/day_07.py
import re
from pathlib import Path


def bag_contains(rules, colour, target) -> bool:
    return target in rules[colour] or any(bag_contains(rules, c, target) for c in rules[colour])


def bag_count(rules, colour) -> int:
    return 1 + sum(q * bag_count(rules, c) for c, q in rules[colour].items())


def main():
    text = (Path(__file__).parent / "../input/input_07.txt").read_text()
    rules = {}
    for line in text.splitlines():
        outer, inner = line.split(" bags contain ")
        rules[outer] = {m[2]: int(m[1]) for m in re.finditer(r"(\d+) (.+?) bag", inner)}
    print("Part 1:", sum(bag_contains(rules, c, "shiny gold") for c in rules))
    print("Part 2:", bag_count(rules, "shiny gold") - 1)


if __name__ == "__main__":
    main()

File: 2020/rs/src/bin/day_07.rs
use std::collections::HashMap;
use std::fs;

fn bag_contains(rules: &HashMap<&str, HashMap<&str, u32>>, colour: &str, target: &str) -> bool {
    rules[colour].contains_key(target)
    || rules[colour].keys().any(|c| bag_contains(rules, c, target))
}

fn bag_count(rules: &HashMap<&str, HashMap<&str, u32>>, colour: &str) -> u32 {
    1 + rules[colour].iter().map(|(c, q)| q * bag_count(rules, c)).sum::<u32>()
}

fn main() {
    let input = fs::read_to_string("../input/input_07.txt").unwrap();
    let rules: HashMap<_, _> = input.lines().map(|line| {
        let key = line.split(" bags contain ").next().unwrap();
        let counts: HashMap<_, _> = line.match_indices(char::is_numeric).map(|(i, s)| {
            let end = i + line[i..].find(" bag").unwrap();
            (&line[i+2..end], s.parse::<u32>().unwrap())
        }).collect();
        (key, counts)
    }).collect();
    println!("Part 1: {}", rules.keys().filter(|c| bag_contains(&rules, c, "shiny gold")).count());
    println!("Part 2: {}", bag_count(&rules, "shiny gold") - 1);
}
